app/models/video.rb

```diff
@@ -0,0 +1,34 @@
+class Video
+  attr_accessor :url
+
+  def initialize(url = nil)
+    self.url = url
+  end
+
+  def twitch?
+    url ? URI.parse(url).host.match?(/^(?:www\.)?(?:twitch\.tv)$/i) : false
+  end
+
+  def youtube?
+    url ? /^(?:https?:\/\/)?(?:www\.)?youtu(?:be\.com|\.be)\/.+/i.match?(url) : false
+  end
+
+  def id
+    return nil unless url
+    if youtube?
+      # https://www.youtube.com/watch?v=asdf1234&feature=related
+      # http://www.youtube.com/embed/watch?feature=player_embedded&v=asdf1234
+      # https://youtu.be/asdf1234
+
+      id = /\/watch.*?v=(.+?)(?:&|\z)/i.match(url)
+      return id[1] if id
+      id = /.+\/([^\/]+$)/.match(url)
+      id ? id[1] : nil
+    elsif twitch?
+      # https://www.twitch.tv/videos/12345678
+
+      id = /.+\/([^\/]+$)/.match(url)
+      id ? id[1] : nil
+    end
+  end
+end
```

Comment:
I agree the complexity of video stuff is now at the point that splitting it into a class is a good idea! I'd go one step further and say this should be its own ActiveRecord model, database table, etc. for full separation. It might only be one or two columns to start, but I think it's worth it for the simpler mental model.

Reply:
Sounds good. Are you thinking you want to go as far as using STI and having a base `Video` class with child classes of `class Twitch < Video` and keeping their specific logic contained therein?

Reply:
My first gut reaction is that STI sounds a bit overkill here, but I'll leave the decision to you and what feels cleanest. It wouldn't be the end of the world if it was still a generic `url` field that we inspect in `Video` methods, since that's the simplest baby step for the data. We can always further refine it later.

Reply:
The current version is now just what you described. `Video` is a simple AR model that acts basically like the PORO version was.

Doing this involved many changes in getting rid of `Run.video_url`, but I think I pretty well got them all and the test suite is still green. Due to `strong_migrations`, there will need to be 1 deploy done before you can drop `video_url` on `runs`, but the column is currently being ignored via ActiveRecord.